#!/bin/bash

awk '{if ($7=="PASS") print $0}' $1 > OK.SNVs
awk '{if ($7=="PASS") print $0}' $2 > OK.indels 
b=(1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M)
c=(1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M)
sed 's/chr//' OK.SNVs > o
sed 's/chr//' OK.indels > oo
for b in ${b[@]}
do
        awk '{if ($1=="'$b'") print $0}' o | sort -k2n > chr$b.SNV.vcf
        awk '{if ($1=="'$b'") print $0}' oo | sort -k2n > chr$b.Indel.vcf
done

for c in ${c[@]}
do

        sg_extraer_SNPs_fichero_tabulado28032011.pl /data/results/Solid0065/info_paneles/cardio/chr$c\_intervalos_sondas chr$c.SNV.vcf  snps > chr$c\_SNVs_N_rango
        sg_extraer_SNPs_fichero_tabulado28032011.pl /data/results/Solid0065/info_paneles/cardio/chr$c\_intervalos_sondas chr$c.Indel.vcf  indels > chr$c\_Indels_N_rango

done

cat chr*_N_rango > variantes_N_rango.vcf

sg_variant_effect_predictor_Ensembl_190411.pl -i variantes_N_rango.vcf  -o variant_effect_output.txt -format vcf -b 1000

rm chr*.SNV.vcf
rm chr*.Indel.vcf
rm chr*_SNVs_N_rango
rm chr*_Indels_N_rango

u=(1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M)
i=(1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M)
a=(1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y M)
for i in ${i[@]}  
do
	awk '{if ($1=="'$i'") print $0}' variantes_N_rango.vcf | sort -k2n > chr$i.vcf
	sed 's/_/\t/' variant_effect_output.txt > ook
	sed 's/_/\t/' ook > ok
	awk '{if ($1=="'$i'") print $0}' ok | sort -k2n > chr$i.annot
done

rm ok
rm ook
for a in ${a[@]}
do
	sg_combinar_vcf_anotacion_APIs.pl chr$a.vcf chr$a.annot > chr$a.comb
	
done



for u in ${u[@]}
do
	sg_extraerBiobaseInformation.pl chr$u.comb /data/results/Solid0065/info_paneles/HGMD-Biobase/bioBaseSortUniques_chr$u.txt > chr$u.annot.biobase
done
cat chr*.annot.biobase  > variantes_anotadas_combinadas.tab_tmp
rm chr*.comb
rm chr*.annot
rm chr*.vcf
rm chr*.annot.biobase
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($11=="ENST00000261201") || ($11=="ENST00000290378") || ($11=="ENST00000366578") || ($11=="ENST00000356239") || ($11=="ENST00000357077") || ($11=="ENST00000277549") || ($11=="ENST00000399655") || ($11=="ENST00000377329") || ($11=="ENST00000269881") || ($11=="ENST00000261448") || ($11=="ENST00000343849") || ($11=="ENST00000265968") || ($11=="ENST00000373960") || ($11=="ENST00000280904") || ($11=="ENST00000261590") || ($11=="ENST00000379802") || ($11=="ENST00000348997") || ($11=="ENST00000367895") || ($11=="ENST00000316623") || ($11=="ENST00000262464") || ($11=="ENST00000223528") || ($11=="ENST00000271348") || ($11=="ENST00000282541") || ($11=="ENST00000372980") || ($11=="ENST00000310706") || ($11=="ENST00000252321") || ($11=="ENST00000399289") || ($11=="ENST00000290310") || ($11=="ENST00000310128") || ($11=="ENST00000262186") || ($11=="ENST00000243457") || ($11=="ENST00000155840") || ($11=="ENST00000200639")  || ($11=="ENST00000372066") || ($11=="ENST00000368301") || ($11=="ENST00000261349") || ($11=="ENST00000354410") || ($11=="ENST00000256993") || ($11=="ENST00000356287") || ($11=="ENST00000361594") || ($11=="ENST00000228841") || ($11=="ENST00000292327") || ($11=="ENST00000375985") || ($11=="ENST00000307128") || ($11=="ENST00000334785") || ($11=="ENST00000376480") || ($11=="ENST00000070846") || ($11=="ENST00000357525") || ($11=="ENST00000287878") || ($11=="ENST00000324501") || ($11=="ENST00000366783") || ($11=="ENST00000369519") || ($11=="ENST00000355794") || ($11=="ENST00000262631") || ($11=="ENST00000299333") || ($11=="ENST00000324727") || ($11=="ENST00000333535") || ($11=="ENST00000337851") || ($11=="ENST00000281456") || ($11=="ENST00000217381") || ($11=="ENST00000299328") || ($11=="ENST00000309889") || ($11=="ENST00000238682") || ($11=="ENST00000295754") || ($11=="ENST00000306077") || ($11=="ENST00000266732") || ($11=="ENST00000232975") || ($11=="ENST00000344887") || ($11=="ENST00000367319") || ($11=="ENST00000403994") || ($11=="ENST00000342992") || ($11=="ENST00000211998") ) print $0;}' variantes_anotadas_combinadas.tab_tmp > variantes_anotadas_combinadas.tab
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000134755") || ($10=="ENSG00000046604") || ($10=="ENSG00000096696") || ($10=="ENSG00000173801") || ($10=="ENSG00000057294") || ($10=="ENSG00000198626") || ($10=="ENSG00000119699") || ($10=="ENSG00000170876") ) print $0; }' variantes_anotadas_combinadas.tab  > Displasia_Arritmogenica_del_Ventriculo_Derecho.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000159251") || ($10=="ENSG00000077522") || ($10=="ENSG00000182533") || ($10=="ENSG00000129170") || ($10=="ENSG00000149596") || ($10=="ENSG00000122367") || ($10=="ENSG00000172399") || ($10=="ENSG00000134571") || ($10=="ENSG00000197616") || ($10=="ENSG00000092054") || ($10=="ENSG00000111245") || ($10=="ENSG00000160808") || ($10=="ENSG00000101306") || ($10=="ENSG00000198523") || ($10=="ENSG00000106617") || ($10=="ENSG00000151729") || ($10=="ENSG00000173991") || ($10=="ENSG00000114854") || ($10=="ENSG00000129991") || ($10=="ENSG00000118194") || ($10=="ENSG00000140416") || ($10=="ENSG00000155657") || ($10=="ENSG00000035403") ) print $0 ;}' variantes_anotadas_combinadas.tab > Miocardiopatia_Hipertrofica.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000069431") || ($10=="ENSG00000159251") || ($10=="ENSG00000077522") || ($10=="ENSG00000175084") || ($10=="ENSG00000046604") || ($10=="ENSG00000134769") || ($10=="ENSG00000112319") || ($10=="ENSG00000106692") || ($10=="ENSG00000122367") || ($10=="ENSG00000160789") || ($10=="ENSG00000134571") || ($10=="ENSG00000197616") || ($10=="ENSG00000092054") || ($10=="ENSG00000162614") || ($10=="ENSG00000198523") || ($10=="ENSG00000080815") || ($10=="ENSG00000143801") || ($10=="ENSG00000203867") || ($10=="ENSG00000183873") || ($10=="ENSG00000129170") || ($10=="ENSG00000170624") || ($10=="ENSG00000102125") || ($10=="ENSG00000173991") || ($10=="ENSG00000120802") || ($10=="ENSG00000114854") || ($10=="ENSG00000129991") || ($10=="ENSG00000118194") || ($10=="ENSG00000140416") || ($10=="ENSG00000155657") || ($10=="ENSG00000035403") ) print $0 ;}' variantes_anotadas_combinadas.tab > Miocardiopatia_Dilatada.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";}{if (($10=="ENSG00000069431") || ($10=="ENSG00000159251") || ($10=="ENSG00000077522") || ($10=="ENSG00000182533") || ($10=="ENSG00000129170") || ($10=="ENSG00000175084") || ($10=="ENSG00000046604") || ($10=="ENSG00000134769") || ($10=="ENSG00000112319") || ($10=="ENSG00000106692") || ($10=="ENSG00000149596") || ($10=="ENSG00000122367") || ($10=="ENSG00000160789") || ($10=="ENSG00000172399") || ($10=="ENSG00000134571") || ($10=="ENSG00000197616") || ($10=="ENSG00000092054") || ($10=="ENSG00000111245") || ($10=="ENSG00000160808") || ($10=="ENSG00000101306") || ($10=="ENSG00000162614") || ($10=="ENSG00000198523") || ($10=="ENSG00000106617") || ($10=="ENSG00000080815") || ($10=="ENSG00000143801") || ($10=="ENSG00000203867") || ($10=="ENSG00000183873") || ($10=="ENSG00000170624") || ($10=="ENSG00000151729") || ($10=="ENSG00000102125") || ($10=="ENSG00000173991") || ($10=="ENSG00000120802") || ($10=="ENSG00000114854") || ($10=="ENSG00000129991") || ($10=="ENSG00000118194") || ($10=="ENSG00000140416") || ($10=="ENSG00000155657") || ($10=="ENSG00000035403")  ) print $0 ;}' variantes_anotadas_combinadas.tab > Miocardiopatia_familiar.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000166147") || ($10=="ENSG00000138829") || ($10=="ENSG00000163513") ) print $0; }' variantes_anotadas_combinadas.tab > Aneurisma_aorta.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000151067") || ($10=="ENSG00000165995") || ($10=="ENSG00000152642") || ($10=="ENSG00000175538") || ($10=="ENSG00000105711") || ($10=="ENSG00000166257") || ($10=="ENSG00000183873") ) print $0 ;}' variantes_anotadas_combinadas.tab > Sindrome_Brugada.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000127914") || ($10=="ENSG00000145362") || ($10=="ENSG00000151067") || ($10=="ENSG00000182533") || ($10=="ENSG00000180509") || ($10=="ENSG00000159197") || ($10=="ENSG00000055118") || ($10=="ENSG00000123700") || ($10=="ENSG00000053918") || ($10=="ENSG00000177098") || ($10=="ENSG00000183873") || ($10=="ENSG00000101400") ) print $0 ;}' variantes_anotadas_combinadas.tab > QT_largo.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000148408") || ($10=="ENSG00000151067") || ($10=="ENSG00000055118") || ($10=="ENSG00000053918") || ($10=="ENSG00000123700")) print $0;}' variantes_anotadas_combinadas.tab > QT_corto.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000143140") || ($10=="ENSG00000130037") || ($10=="ENSG00000159197") || ($10=="ENSG00000053918") || ($10=="ENSG00000175206")) print $0 ;}' variantes_anotadas_combinadas.tab > Fibrilacion_auricular.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000118729") || ($10=="ENSG00000198626") ) print $0 ;}' variantes_anotadas_combinadas.tab > Polimorfica_Catecolaminergica.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000127914") || ($10=="ENSG00000145362") || ($10=="ENSG00000151067") || ($10=="ENSG00000165995") || ($10=="ENSG00000118729") || ($10=="ENSG00000182533") || ($10=="ENSG00000134755") || ($10=="ENSG00000046604") || ($10=="ENSG00000096696") || ($10=="ENSG00000152642") || ($10=="ENSG00000173801") || ($10=="ENSG00000130037") || ($10=="ENSG00000180509") || ($10=="ENSG00000159197") || ($10=="ENSG00000175538") || ($10=="ENSG00000055118") || ($10=="ENSG00000123700") || ($10=="ENSG00000053918") || ($10=="ENSG00000175206") || ($10=="ENSG00000057294")  || ($10=="ENSG00000198523") || ($10=="ENSG00000198626") || ($10=="ENSG00000105711") || ($10=="ENSG00000166257") || ($10=="ENSG00000177098") || ($10=="ENSG00000183873") || ($10=="ENSG00000101400") || ($10=="ENSG00000119699") || ($10=="ENSG00000170876") ) print $0;}' variantes_anotadas_combinadas.tab > Arritmia_familiar.xls
awk 'BEGIN {print "Gene_name\tChr\tPosition\tReference_genotype\tSample_genotype\tVariation_type\tVariation_length\tZygosity\tVariant_ID\tEnsembl_Gene_ID\tEnsembl_Transcript_ID\tGene_description\tConsequence_on_transcripts\tcDNA_position\taa_position\taa_change\tGERP_conservarion_score\tCoverage\tGenotype_quality\tMapping_quality\tDisease\tHGVS_ID\tPubMed_ID";} {if (($10=="ENSG00000069431") || ($10=="ENSG00000159251") || ($10=="ENSG00000077522") || ($10=="ENSG00000127914") || ($10=="ENSG00000145362") || ($10=="ENSG00000148408") || ($10=="ENSG00000151067") || ($10=="ENSG00000165995") || ($10=="ENSG00000141979") || ($10=="ENSG00000118729") || ($10=="ENSG00000182533") || ($10=="ENSG00000129170") || ($10=="ENSG00000175084") || ($10=="ENSG00000134755") || ($10=="ENSG00000046604") || ($10=="ENSG00000096696") || ($10=="ENSG00000134769") || ($10=="ENSG00000112319") || ($10=="ENSG00000166147") || ($10=="ENSG00000138829") || ($10=="ENSG00000106692") || ($10=="ENSG00000143140") || ($10=="ENSG00000152642") || ($10=="ENSG00000149596") || ($10=="ENSG00000173801") || ($10=="ENSG00000130037") || ($10=="ENSG00000180509") || ($10=="ENSG00000159197") || ($10=="ENSG00000175538") || ($10=="ENSG00000055118") || ($10=="ENSG00000123700") || ($10=="ENSG00000053918") || ($10=="ENSG00000005893") || ($10=="ENSG00000122367") || ($10=="ENSG00000160789") || ($10=="ENSG00000070018") || ($10=="ENSG00000068305") || ($10=="ENSG00000134571") || ($10=="ENSG00000197616") || ($10=="ENSG00000092054") || ($10=="ENSG00000111245") || ($10=="ENSG00000160808") || ($10=="ENSG00000101306") || ($10=="ENSG00000172399") || ($10=="ENSG00000162614") || ($10=="ENSG00000175206") || ($10=="ENSG00000057294") || ($10=="ENSG00000198523") || ($10=="ENSG00000106617") || ($10=="ENSG00000080815") || ($10=="ENSG00000143801") || ($10=="ENSG00000203867") || ($10=="ENSG00000198626") || ($10=="ENSG00000105711") || ($10=="ENSG00000166257") || ($10=="ENSG00000177098") || ($10=="ENSG00000183873") || ($10=="ENSG00000170624") || ($10=="ENSG00000151729") || ($10=="ENSG00000101400") || ($10=="ENSG00000102125") || ($10=="ENSG00000173991") || ($10=="ENSG00000119699") || ($10=="ENSG00000163513") || ($10=="ENSG00000170876") || ($10=="ENSG00000120802") || ($10=="ENSG00000114854") || ($10=="ENSG00000129991") || ($10=="ENSG00000118194") || ($10=="ENSG00000140416") || ($10=="ENSG00000155657") || ($10=="ENSG00000035403")) print $0;}'  variantes_anotadas_combinadas.tab  > muerte_subita.xls
rm variantes_anotadas_combinadas.tab_tmp 
